experimental/entity/entity.proto

```diff
@@ -0,0 +1,219 @@
+syntax = "proto3";
+package entity;
+
+option go_package = "./;entity";
+
+message EntityMessage {
+    // Path encodes the UID and kind.  It will include / to indicate folders
+    string path = 1; 
+
+    // The kind (extracted from path suffix)
+    string kind = 2;
+
+    // The raw entity bytes
+    bytes payload = 3; 
+
+    // Metadata managed by the underlying storage engine(s)
+    StorageMetadata meta = 4; 
+}
+
+// System manages these properties
+message StorageMetadata {
+    // Unix millis for the creation time
+    int64 createdAt = 1;
+
+    // Unix millis for last update time
+    int64 updatedAt = 2;
+
+    // UserID that last updated the entity
+    string updatedBy = 3;
+
+    // Version identifier. This may include additional location information (branch/revision/etc)
+    // The version string is not sortable with other versions
+    string version = 4; 
+
+    // Content checksum
+    string etag = 5; 
+
+    // Size in bytes
+    int64 size = 6;
+
+    // list of operations the calling user can perform
+    repeated string allowedOperations = 7; 
+
+    // avaliable PRs for this path
+    repeated EntityPR prs = 8; 
+}
+
+message EntityPR {
+    // PR Title
+    string name = 1;
+
+    // passed to get and it will load the content
+    string version = 2; 
+
+    // Unix millis for last update time
+    int64 updatedAt = 3;
+
+    // External URL
+    string url = 4;
+}
+
+message FolderListing {
+    string path = 1; 
+
+    // Child entity information
+    repeated EntityMessage items = 2;
+
+    // More items exist
+    string nextPageToken = 3;
+}
+
```

Comment:
Will this not be very slow to implement behind file system / blob API? The list operations for file systems / S3 cannot return the items body (so nothing of our envelope)

Reply:
We need to figure out the best way to implement -- I suspect using search or creating index files gives us some good options.

GCS allows custom metadata, that you can then use in search/listing:
https://cloud.google.com/storage/docs/listing-objects#prereq-rest

<img width="555" alt="image" src="https://user-images.githubusercontent.com/705951/188030846-46727ebf-a167-4431-92af-cae33816b311.png">

S3 allows setting custom metadata, and will return it in a list, *BUT* does not lets you filter by metadata in their list APIs

In SQL, we can obviously do whatever we need ;)

Reply:
In GCS, any metadata is returned in the list call:
<img width="838" alt="image" src="https://user-images.githubusercontent.com/705951/188034691-72cbec94-7a60-413d-adce-1eb8078b661f.png">

Reply:
so custom meta data can be returned, but not body. The current FolderListing response contains EntityMessage items (which include payload body). Will FolderListing contain items without payload? Maybe should be FolderListingItem and not full EntityMessage.

Reply:
@sdboyer agree relying on fs metadata feels a bit scary here, but what other option do we have if we want actually useful info in list operations?  (Other than having a local db representation of it that we can use for list operations and never do lists directly via external blob store)

Reply:
I think it's about metadata as defined here: https://github.com/grafana/grafana-plugin-sdk-go/pull/527/files/5e31f52bc7832bfc5ee4611c0ecd4157af395eb1#diff-48588ccb7f56939392159a401f922d5484b9a3523dd2ba98e86d6792785cab94R21-R46 - it seems reasonable to use the "native fs metadata" for this kind of "raw file meta"

Reply:
relaying on metadata implementations in S3 or GCS feels dangerous. We want to be able to deploy this in all cloud providers. I would rather run SQL everywhere than having some on GCS and some on SQL.

Reply:
@ArturWierzbicki yeah the gap between raw and non-raw types here is significant.

However, if what you linked is what's being represented, then...i'm confused, because `who` is not one of the fields in that proto message :)

i'm also inclined to agree with @bergquist, variations in filesystem metadata feel too scary to rely on for recording this information (as well as being wholly insufficient for the non-raw cases i have in mind)

Reply:
@torkelo

> (Other than having a local db representation of it that we can use for list operations and never do lists directly via external blob store)

And...i mean, to be blunt, but i think that _is_ the only option.

Which is to say, my immediate gut reaction to seeing this, and considering the potential variation in metadata implementations by different backends not being something we can reliably abstract across, and the likely insufficiency for non-raw metadata cases, as a serious mark against an fs-backed approach as a whole.

Not saying my mind's made up, just offering my immediate reaction

Reply:
> @ArturWierzbicki yeah the gap between raw and non-raw types here is significant.
> 
> However, if what you linked is what's being represented, then...i'm confused, because `who` (in the image) is not one of the fields in that proto message :)
> 
> i'm also inclined to agree with @bergquist, variations in filesystem metadata feel too scary to rely on for recording this information (as well as being wholly insufficient for the non-raw cases i have in mind)

I assume the `who` can be mapped to `updatedBy` :) I do not think its intended to satisfy the use case of the elaborate-config-type metadata - that's going to stay inside the file/payload. 

The metadata in question is the basic information about file such as creation date, creator, size, etc. Note that this is useful for both raw and "confgi" file types